File: src/engine/similarity.ts
import { cosine } from '@/lib/math';
import { EVENTS, EVENT_TAGS, MACRO_CONTEXT, EventDef, MacroContext } from '@/config/events';
import { ANALOGUE_WEIGHTS, TRIGGER_ZSCORE_SIGMA, SIMILARITY_ASSET_POOL } from '@/config/engine';
import { poiRet, EventReturns } from './returns';

export interface AnalogueScore {
  event: string;
  composite: number;
  rawComposite: number;
  quant: number;
  quant_pt: number;
  tag: number;
  macro: number;
  sharedAssetCount: number;
  coverageRatio: number;
  sparsePenalty: number;
  confidenceLabel: 'high' | 'medium' | 'thin';
}

interface RunAnalogueMatchOptions {
  weights?: typeof ANALOGUE_WEIGHTS;
  simAssets?: string[];
  events?: EventDef[];
  eventTags?: Record<string, Set<string>>;
  macroContext?: Record<string, MacroContext>;
}

const MIN_SHARED_ASSETS = 3;
const NORMAL_COVERAGE_RATIO = 0.5;

export function coveragePenalty(sharedAssetCount: number, requestedAssetCount: number): {
  coverageRatio: number;
  sparsePenalty: number;
  confidenceLabel: 'high' | 'medium' | 'thin';
} {
  if (requestedAssetCount <= 0) {
    return { coverageRatio: 0, sparsePenalty: 0.35, confidenceLabel: 'thin' };
  }

  const coverageRatio = sharedAssetCount / requestedAssetCount;
  const ratioPenalty = Math.max(0.35, Math.min(1, coverageRatio / NORMAL_COVERAGE_RATIO));
  const floorPenalty = sharedAssetCount >= MIN_SHARED_ASSETS
    ? 1
    : Math.max(0.25, sharedAssetCount / MIN_SHARED_ASSETS);
  const sparsePenalty = Math.min(1, ratioPenalty * floorPenalty);

  let confidenceLabel: 'high' | 'medium' | 'thin' = 'thin';
  if (sharedAssetCount >= Math.max(MIN_SHARED_ASSETS + 2, Math.ceil(requestedAssetCount * 0.75))) {
    confidenceLabel = 'high';
  } else if (sharedAssetCount >= MIN_SHARED_ASSETS && coverageRatio >= 0.4) {
    confidenceLabel = 'medium';
  }

  return { coverageRatio, sparsePenalty, confidenceLabel };
}

export function tagSim(a: Set<string>, b: Set<string>): number {
  if (a.size === 0 && b.size === 0) return 1;
  const union = new Set([...a, ...b]);
  const intersection = [...a].filter((value) => b.has(value));
  return union.size > 0 ? intersection.length / union.size : 0;
}

export function macroSim(
  liveTriggerZ: number | null,
  liveCpi: string,
  liveFed: string,
  historicalContext: { trigger?: number; cpi?: string; fed?: string },
  historicalTriggerZ: number | null,
): number {
  let score = 0;

  if (liveTriggerZ !== null && historicalTriggerZ !== null) {
    const sigma = TRIGGER_ZSCORE_SIGMA;
    const triggerScore = Math.exp(-0.5 * ((liveTriggerZ - historicalTriggerZ) / sigma) ** 2);
    score += triggerScore * 0.4;
  }

  if (liveCpi === (historicalContext.cpi || '')) score += 0.35;
  if (liveFed === (historicalContext.fed || '')) score += 0.25;

  return score;
}

function nearestValueAtOrBefore(
  series: Record<number, number> | undefined,
  targetOffset: number,
  tolerance: number,
): number {
  if (!series) return Number.NaN;
  const offsets = Object.keys(series)
    .map(Number)
    .filter((offset) => Math.abs(offset - targetOffset) <= tolerance)
    .sort((left, right) => left - right);
  if (offsets.length === 0) return Number.NaN;
  const below = offsets.filter((offset) => offset <= targetOffset);
  const bestOffset = below.length > 0 ? below[below.length - 1] : offsets[0];
  return series[bestOffset];
}

function pathVec(
  returnsByAsset: Record<string, Record<number, number>>,
  assets: string[],
  dayN: number,
): number[] {
  const offsets = Array.from({ length: dayN + 1 }, (_, index) => index);
  const vector: number[] = [];

  for (const asset of assets) {
    const series = returnsByAsset[asset];
    for (const offset of offsets) {
      const value = nearestValueAtOrBefore(series, offset, 1);
      vector.push(Number.isNaN(value) ? 0 : value);
    }
  }

  return vector;
}

export function runAnalogueMatch(
  eventReturns: EventReturns,
  liveReturns: Record<string, Record<number, number>>,
  liveTags: Set<string>,
  liveTriggerZ: number | null,
  liveCpi: string,
  liveFed: string,
  dayN: number,
  triggerZScores: Record<string, number>,
  options: RunAnalogueMatchOptions = {},
): AnalogueScore[] {
  const events = options.events || EVENTS;
  const eventTags = options.eventTags || EVENT_TAGS;
  const macroContext = options.macroContext || MACRO_CONTEXT;
  const simAssets = options.simAssets || SIMILARITY_ASSET_POOL;
  const weights = options.weights || ANALOGUE_WEIGHTS;
  const livePool = simAssets.filter((asset) => liveReturns[asset] && Object.keys(liveReturns[asset]).length > 0);
  const availableLiveOffsets = new Set<number>();

  for (const asset of livePool) {
    for (const offset of Object.keys(liveReturns[asset]).map(Number)) {
      availableLiveOffsets.add(offset);
    }
  }

  const scoringDayN = availableLiveOffsets.size > 0
    ? Math.min(dayN, Math.max(...Array.from(availableLiveOffsets)))
    : dayN;

  const livePointVec = simAssets.map((asset) => nearestValueAtOrBefore(liveReturns[asset], scoringDayN, 2));
  const weightSum = weights.quant + weights.tag + weights.macro;
  const normalizedWeights = weightSum > 0
    ? {
        quant: weights.quant / weightSum,
        tag: weights.tag / weightSum,
        macro: weights.macro / weightSum,
      }
    : { ...ANALOGUE_WEIGHTS };

  const scores: AnalogueScore[] = [];

  for (const event of events) {
    const eventName = event.name;
    const histPointVec = simAssets.map((asset) => poiRet(eventReturns, asset, eventName, scoringDayN));
    const quantPoint = (cosine(livePointVec, histPointVec) + 1) / 2;

    const sharedAssets = livePool.filter((asset) => {
      const hist = eventReturns[asset]?.[eventName];
      return hist && Object.keys(hist).length > 0;
    });

    const livePathVec = pathVec(liveReturns, simAssets, scoringDayN);
    const historicalPathDict: Record<string, Record<number, number>> = {};
    for (const asset of simAssets) {
      const series = eventReturns[asset]?.[eventName];
      if (series) {
        historicalPathDict[asset] = series;
      }
    }
    const histPathVec = pathVec(historicalPathDict, simAssets, scoringDayN);
    const quantPath = (cosine(livePathVec, histPathVec) + 1) / 2;

    const tag = tagSim(liveTags, eventTags[eventName] || new Set());
    const macro = macroSim(
      liveTriggerZ,
      liveCpi,
      liveFed,
      macroContext[eventName] || {},
      triggerZScores[eventName] ?? null,
    );

    const rawComposite =
      normalizedWeights.quant * quantPath +
      normalizedWeights.tag * tag +
      normalizedWeights.macro * macro;
    const coverage = coveragePenalty(sharedAssets.length, livePool.length);
    const composite = rawComposite * coverage.sparsePenalty;

    scores.push({
      event: eventName,
      composite,
      rawComposite,
      quant: quantPath,
      quant_pt: quantPoint,
      tag,
      macro,
      sharedAssetCount: sharedAssets.length,
      coverageRatio: coverage.coverageRatio,
      sparsePenalty: coverage.sparsePenalty,
      confidenceLabel: coverage.confidenceLabel,
    });
  }

  scores.sort((left, right) => right.composite - left.composite);
  return scores;
}

export function selectEvents(scores: AnalogueScore[], cutoff: number): string[] {
  const selected = scores.filter((score) => score.composite >= cutoff).map((score) => score.event);
  return selected.length > 0 ? selected : scores.map((score) => score.event);
}

export function filterScoresByActiveEvents(
  scores: AnalogueScore[],
  activeEvents: Set<string>,
): AnalogueScore[] {
  return scores.filter((score) => activeEvents.has(score.event));
}

export function compositeReturn(
  eventReturns: EventReturns,
  label: string,
  selectedEvents: string[],
  scores: AnalogueScore[]
): Record<number, number> | null {
  const scoreMap = new Map(scores.map((score) => [score.event, score.composite]));
  const weightedValues: Record<number, number> = {};
  const weightSums: Record<number, number> = {};

  for (const eventName of selectedEvents) {
    const series = eventReturns[label]?.[eventName];
    if (!series) continue;
    const weight = scoreMap.get(eventName) || 0;
    for (const [offsetStr, value] of Object.entries(series)) {
      const offset = parseInt(offsetStr, 10);
      weightedValues[offset] = (weightedValues[offset] || 0) + value * weight;
      weightSums[offset] = (weightSums[offset] || 0) + weight;
    }
  }

  const result: Record<number, number> = {};
  for (const offset of Object.keys(weightedValues).map(Number)) {
    if (weightSums[offset] > 0) {
      result[offset] = weightedValues[offset] / weightSums[offset];
    }
  }

  return Object.keys(result).length > 0 ? result : null;
}

File: engine/similarity.py
from __future__ import annotations

import math

from config.defaults import ANALOGUE_WEIGHTS, SIMILARITY_ASSET_POOL, TRIGGER_ZSCORE_SIGMA
from config.events import EVENT_TAGS, EVENTS, MACRO_CONTEXT
from engine.math_utils import cosine
from engine.models import AnalogueScore, EventDef, EventReturns, MacroContext
from engine.returns import poi_ret


MIN_SHARED_ASSETS = 3
NORMAL_COVERAGE_RATIO = 0.5


def coverage_penalty(shared_asset_count: int, requested_asset_count: int) -> tuple[float, float, str]:
    if requested_asset_count <= 0:
        return 0.0, 0.35, "thin"
    coverage_ratio = shared_asset_count / requested_asset_count
    ratio_penalty = max(0.35, min(1.0, coverage_ratio / NORMAL_COVERAGE_RATIO))
    floor_penalty = 1.0 if shared_asset_count >= MIN_SHARED_ASSETS else max(0.25, shared_asset_count / MIN_SHARED_ASSETS)
    sparse_penalty = min(1.0, ratio_penalty * floor_penalty)
    confidence = "thin"
    if shared_asset_count >= max(MIN_SHARED_ASSETS + 2, math.ceil(requested_asset_count * 0.75)):
        confidence = "high"
    elif shared_asset_count >= MIN_SHARED_ASSETS and coverage_ratio >= 0.4:
        confidence = "medium"
    return coverage_ratio, sparse_penalty, confidence


def tag_sim(left: set[str], right: set[str]) -> float:
    if not left and not right:
        return 1.0
    union = left | right
    return len(left & right) / len(union) if union else 0.0


def macro_sim(
    live_trigger_z: float | None,
    live_cpi: str,
    live_fed: str,
    historical_context: MacroContext | dict,
    historical_trigger_z: float | None,
) -> float:
    score = 0.0
    historical_cpi = historical_context.cpi if isinstance(historical_context, MacroContext) else historical_context.get("cpi", "")
    historical_fed = historical_context.fed if isinstance(historical_context, MacroContext) else historical_context.get("fed", "")
    if live_trigger_z is not None and historical_trigger_z is not None:
        trigger_score = math.exp(-0.5 * ((live_trigger_z - historical_trigger_z) / TRIGGER_ZSCORE_SIGMA) ** 2)
        score += trigger_score * 0.4
    if live_cpi == historical_cpi:
        score += 0.35
    if live_fed == historical_fed:
        score += 0.25
    return score


def nearest_value_at_or_before(series: dict[int, float] | None, target_offset: int, tolerance: int) -> float:
    if not series:
        return math.nan
    offsets = sorted(offset for offset in series if abs(offset - target_offset) <= tolerance)
    if not offsets:
        return math.nan
    below = [offset for offset in offsets if offset <= target_offset]
    best = below[-1] if below else offsets[0]
    return series[best]


def path_vec(returns_by_asset: dict[str, dict[int, float]], assets: list[str], day_n: int) -> list[float]:
    vector: list[float] = []
    for asset in assets:
        series = returns_by_asset.get(asset)
        for offset in range(day_n + 1):
            value = nearest_value_at_or_before(series, offset, 1)
            vector.append(0.0 if math.isnan(value) else value)
    return vector


def run_analogue_match(
    event_returns: EventReturns,
    live_returns: dict[str, dict[int, float]],
    live_tags: set[str],
    live_trigger_z: float | None,
    live_cpi: str,
    live_fed: str,
    day_n: int,
    trigger_z_scores: dict[str, float],
    *,
    weights: dict[str, float] | None = None,
    sim_assets: list[str] | None = None,
    events: list[EventDef] | None = None,
    event_tags: dict[str, set[str]] | None = None,
    macro_context: dict[str, MacroContext] | None = None,
) -> list[AnalogueScore]:
    events = events or EVENTS
    event_tags = event_tags or EVENT_TAGS
    macro_context = macro_context or MACRO_CONTEXT
    sim_assets = sim_assets or SIMILARITY_ASSET_POOL
    weights = weights or ANALOGUE_WEIGHTS

    live_pool = [asset for asset in sim_assets if live_returns.get(asset)]
    available_offsets = {offset for asset in live_pool for offset in live_returns.get(asset, {})}
    scoring_day_n = min(day_n, max(available_offsets)) if available_offsets else day_n
    live_point_vec = [nearest_value_at_or_before(live_returns.get(asset), scoring_day_n, 2) for asset in sim_assets]

    weight_sum = sum(weights.values())
    normalized = {
        key: (value / weight_sum if weight_sum > 0 else ANALOGUE_WEIGHTS[key])
        for key, value in weights.items()
    }

    scores: list[AnalogueScore] = []
    for event in events:
        event_name = event.name
        hist_point_vec = [poi_ret(event_returns, asset, event_name, scoring_day_n) for asset in sim_assets]
        quant_point = (cosine(live_point_vec, hist_point_vec) + 1) / 2

        shared_assets = [
            asset
            for asset in live_pool
            if event_returns.get(asset, {}).get(event_name)
        ]
        live_path_vec = path_vec(live_returns, sim_assets, scoring_day_n)
        hist_path_dict = {
            asset: event_returns.get(asset, {}).get(event_name, {})
            for asset in sim_assets
            if event_returns.get(asset, {}).get(event_name)
        }
        hist_path_vec = path_vec(hist_path_dict, sim_assets, scoring_day_n)
        quant_path = (cosine(live_path_vec, hist_path_vec) + 1) / 2
        tag = tag_sim(live_tags, event_tags.get(event_name, set()))
        macro = macro_sim(
            live_trigger_z,
            live_cpi,
            live_fed,
            macro_context.get(event_name, {}),
            trigger_z_scores.get(event_name),
        )
        raw = normalized["quant"] * quant_path + normalized["tag"] * tag + normalized["macro"] * macro
        coverage_ratio, sparse_penalty, confidence = coverage_penalty(len(shared_assets), len(live_pool))
        scores.append(
            AnalogueScore(
                event=event_name,
                composite=raw * sparse_penalty,
                raw_composite=raw,
                quant=quant_path,
                quant_pt=quant_point,
                tag=tag,
                macro=macro,
                shared_asset_count=len(shared_assets),
                coverage_ratio=coverage_ratio,
                sparse_penalty=sparse_penalty,
                confidence_label=confidence,  # type: ignore[arg-type]
            )
        )
    return sorted(scores, key=lambda item: item.composite, reverse=True)


def select_events(scores: list[AnalogueScore], cutoff: float) -> list[str]:
    selected = [score.event for score in scores if score.composite >= cutoff]
    return selected if selected else [score.event for score in scores]


def filter_scores_by_active_events(scores: list[AnalogueScore], active_events: set[str]) -> list[AnalogueScore]:
    return [score for score in scores if score.event in active_events]


def composite_return(
    event_returns: EventReturns,
    label: str,
    selected_events: list[str],
    scores: list[AnalogueScore],
) -> dict[int, float] | None:
    score_map = {score.event: score.composite for score in scores}
    weighted_values: dict[int, float] = {}
    weight_sums: dict[int, float] = {}
    for event_name in selected_events:
        series = event_returns.get(label, {}).get(event_name, {})
        weight = score_map.get(event_name, 0.0)
        for offset, value in series.items():
            weighted_values[offset] = weighted_values.get(offset, 0.0) + value * weight
            weight_sums[offset] = weight_sums.get(offset, 0.0) + weight
    result = {
        offset: weighted_values[offset] / weight_sums[offset]
        for offset in weighted_values
        if weight_sums[offset] > 0
    }
    return result or None
